File: src/main/java/year2020/Day02.java
package year2020;

import aoc.IAocTask;

import java.util.List;

public class Day02 implements IAocTask {
    @Override
    public String getFileName() {
        return "aoc2020/input_02.txt";
    }

    @Override
    public void solvePartOne(List<String> lines) {
        System.out.println(lines.stream()
                .map(PasswordEntry::parse)
                .filter(PasswordEntry::check)
                .count());
    }

    @Override
    public void solvePartTwo(List<String> lines) {
        System.out.println(lines.stream()
                .map(PasswordEntry::parse)
                .filter(PasswordEntry::check2)
                .count());
    }

    static class PasswordEntry {
        int minCount;
        int maxCount;
        char letter;
        String password;

        public PasswordEntry(int minCount, int maxCount, char letter, String password) {
            this.minCount = minCount;
            this.maxCount = maxCount;
            this.letter = letter;
            this.password = password;
        }

        static PasswordEntry parse(String line) {
            String[] limitsLetterPassword = line.split(" ");
            String[] minMax = limitsLetterPassword[0].split("-");
            int minCount = Integer.parseInt(minMax[0]);
            int maxCount = Integer.parseInt(minMax[1]);
            char letter = limitsLetterPassword[1].replace(":", "").charAt(0);
            String password = limitsLetterPassword[2];
            return new PasswordEntry(minCount, maxCount, letter, password);
        }

        boolean check() {
            int count = 0;
            for (int i = 0; i < password.length(); i++) {
                if (password.charAt(i) == letter) {
                    count++;
                }
            }
            return minCount <= count && count <= maxCount;
        }

        boolean check2() {
            return password.charAt(minCount - 1) == letter ^ password.charAt(maxCount - 1) == letter;
        }
    }
}

File: src/main/kotlin/year2024/Day02.kt
package year2024

import aoc.IAocTaskKt


class Day02 : IAocTaskKt {
    override fun getFileName(): String = "aoc2024/input_02.txt"

    override fun solvePartOne(lines: List<String>) {
        lines.map { line ->
            line.split(" ")
                .map { it.toInt() }
        }.count { it.isSafe() }
            .let { println(it) }
    }

    override fun solvePartTwo(lines: List<String>) {
        lines.map { line ->
            line.split(" ")
                .map { it.toInt() }
        }.count { it.isSafeWithDampener() }
            .let { println(it) }
    }
}

private fun List<Int>.isSafe(): Boolean = mapIndexed { idx, value -> idx == 0 || this[idx - 1] in (value + 1)..(value + 3) }
        .all { it }
            || mapIndexed { idx, value -> idx == 0 || this[idx - 1] in (value - 3) until value }
        .all { it }

private fun List<Int>.isSafeWithDampener(): Boolean = List(size) { deletedIdx ->
    filterIndexed { idx, _ -> idx != deletedIdx }
        .isSafe()
}.any { it }
